crates/via-router/src/iter.rs

```diff
@@ -1,147 +1,247 @@
-use std::{iter::Peekable, ops::Deref, str::CharIndices};
+use smallvec::SmallVec;
+use std::{iter::Peekable, rc::Rc, str::CharIndices};
 
 use crate::node::{Node, Pattern};
-
-#[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub struct Component<'a, 'b, T> {
+use crate::Store;
+
+/// Represents either a partial or exact match for a given path segment.
+#[derive(Clone, Copy)]
+pub struct Match<'a, 'b, T> {
+    /// Indicates whether or not the match is considered an exact match.
+    /// If the match is exact, both the middleware and responders will be
+    /// called during a request. Otherwise, only the middleware will be
+    /// called.
     pub is_exact_match: bool,
-    pub pattern: Pattern,
-    pub param: Option<(&'static str, &'b str)>,
-    pub route: &'a T,
+
+    /// The value of the path segment that matches self.pattern(). If the
+    /// matched route has a CatchAll pattern, this will be the remainder
+    /// of the url path without the leading `/` character.
+    pub path_segment: &'b str,
+
+    /// The node that matches `self.value`.
+    node: &'a Node<T>,
 }
 
-#[derive(Debug)]
+/// An iterator that yields all possible partial and exact matches for a url path.
 pub struct Visit<'a, 'b, T> {
     node: &'a Node<T>,
-    path: Path<'b>,
-    root: bool,
+    store: &'a Store<T>,
+    depth: usize,
+    index: usize,
+    path_value: &'b str,
+    // TODO: refactor path_segments to be Option<_> and remove the Rc<_> wrapper.
```

Comment:
After running some benchmarks, it turns out that the implementation with `Rc` is faster than the Option wrapper. This is likely due to the value of `path_segments` moving each time the Visit struct is forked.